TLS 1.3: SRV: Improve key exchange mode selection

ronald-cron-arm opened the pull request:
## Description
Fix #8448  

## PR checklist
- [x] **changelog** not required, implementation rework
- [x] **backport** not required, no TLS 1.3 support in 2.28
- [x] **tests** one additional ssl-opt.sh test

yanesca reviewed (approved):
I have a couple of minor comments (no blockers), looks good to me.

ronald-cron-arm commented on library/ssl_tls13_server.c:
```diff
@@ -138,28 +139,36 @@ static int ssl_tls13_offered_psks_check_identity_match_ticket(
+    ret = ssl->conf->f_ticket_parse(ssl->conf->p_ticket,
+                                    session,
+                                    ticket_buffer, identity_len);
+    if (ret == 0) {
```
Yes thanks for the proposal, this is more readable with a switch.

ronald-cron-arm commented on library/ssl_tls13_server.c:
```diff
@@ -138,28 +139,36 @@ static int ssl_tls13_offered_psks_check_identity_match_ticket(
     }
 
-    if (ret != 0) {
+    ret = SSL_TLS1_3_PSK_IDENTITY_MATCH_BUT_PSK_NOT_USABLE;
```
I prefer to keep it here rather that repeat it four times. I have added a comment. Factorizing this affectation does not probably make any difference for a suitably smart compiler but I have no idea if actually it would be just a few of them or all of them.

ronald-cron-arm commented on library/ssl_tls13_server.c:
```diff
@@ -186,11 +185,12 @@ static int ssl_tls13_offered_psks_check_identity_match_ticket(
 
-    ret = MBEDTLS_ERR_SSL_SESSION_TICKET_EXPIRED;
 #if defined(MBEDTLS_HAVE_TIME)
+    ret = MBEDTLS_ERR_SSL_SESSION_TICKET_EXPIRED;
```
This part of the code has be refactored. This does not seem to be relevant anymore.

ronald-cron-arm commented:
CI green, ABI-API-checking failure not related to this PR.

yanesca reviewed (dismissed):
LGTM

tom-cosgrove-arm commented on library/ssl_tls13_server.c:
```diff
@@ -39,6 +39,62 @@ static const mbedtls_ssl_ciphersuite_t *ssl_tls13_validate_peer_ciphersuite(
+
+        /*
+         * If a valid PSK ciphersuite identifier has been passed in, we seek
+         * for an exact match.
```
(nit; ignore if no other changes needed) does "seek for" just mean "want"

tom-cosgrove-arm commented on library/ssl_tls13_server.c:
```diff
@@ -39,6 +39,62 @@ static const mbedtls_ssl_ciphersuite_t *ssl_tls13_validate_peer_ciphersuite(
+        return;
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG(2, ("No matched ciphersuite"));
```
(nit; again, only make changes if other changes needed) Will the debug log already say what we were trying to match?

tom-cosgrove-arm commented on library/ssl_tls13_server.c:
```diff
@@ -90,8 +146,30 @@ static int ssl_tls13_parse_key_exchange_modes_ext(mbedtls_ssl_context *ssl,
+ *
+ * The return value SSL_TLS1_3_PSK_IDENTITY_MATCH_BUT_PSK_NOT_USABLE means
+ * that the pre-shared-key identity matches that of a ticket or an external
+ * provisioned pre-shared-key. We have thus been able to retrieve the
```
(nit; not necessary to change) should it be "externally-provisioned" rather than "external provisioned"?

tom-cosgrove-arm commented on library/ssl_tls13_server.c:
```diff
@@ -138,58 +215,47 @@ static int ssl_tls13_offered_psks_check_identity_match_ticket(
     }
```
(nit; pre-existing) the debug message "buffer too small" seems wrong for `calloc()` failure

ronald-cron-arm replied:
I've just removed it as we usually do not log memory allocation failures.

tom-cosgrove-arm commented on library/ssl_tls13_server.c:
```diff
@@ -481,6 +473,13 @@ static int ssl_tls13_session_copy_ticket(mbedtls_ssl_session *dst,
+    int key_exchange_mode;
+    const mbedtls_ssl_ciphersuite_t *ciphersuite_info;
+};
+#define PSK_ATTRIBUTES_INIT { 0, 0, 0 }
```
(nit) should this be `{ 0, 0, NULL }`?

ronald-cron-arm replied:
Yes better, thanks for catching this.

tom-cosgrove-arm commented on library/ssl_tls13_server.c:
```diff
@@ -1734,10 +1673,11 @@ static int ssl_tls13_parse_client_hello(mbedtls_ssl_context *ssl,
     /* Update checksum with either
      * - The entire content of the CH message, if no PSK extension is present
      * - The content up to but excluding the PSK extension, if present.
+     * Always parse the pre-shared key extension when present in the
```
(nit) everywhere else in this PR this has been written `pre-shared-key`

tom-cosgrove-arm reviewed (dismissed):
LGTM.

Only very minor comments - it's up to you (and @yanesca) if you think any are worth another round of review - I am happy with this as it is

yanesca reviewed (approved):
LGTM

tom-cosgrove-arm commented:
I don't understand why the CI failed after these changes? Something spurious?

```
[2024-03-07T14:49:52.504Z] Connection ID, no packing: Cli+Srv enabled, CID on renegotiation ....... PASS
[2024-03-07T14:50:00.627Z] Connection ID, 3D+MTU: Cli+Srv enabled, CID on renegotiation ........... FAIL
[2024-03-07T14:50:00.627Z]   ! bad client exit code (expected 0, got 1)
[2024-03-07T14:50:00.627Z]   ! outputs saved to o-XXX-128.log
[2024-03-07T14:50:03.231Z] Connection ID: Cli+Srv enabled, Cli disables on renegotiation .......... PASS
```

gilles-peskine-arm commented:
@tom-cosgrove-arm Probably an instance of https://github.com/Mbed-TLS/mbedtls/issues/8471

ronald-cron-arm commented:
@yanesca @tom-cosgrove-arm There was one ssl-opt.sh test failing when merging with development HEAD (merge queue CI). I've rebased on top of development HEAD (trivial rebase, before the rebase the head of this PR was 19a6bc6d9c7727e0b8ccfc0f4fd5962fe23ac714) and added a commit to fix the issue with the ssl-opt.sh test.

tom-cosgrove-arm reviewed (approved):
I've reviewed the rebase and the final commit, and happy with them on top of my earlier approval

ronald-cron-arm commented:
The internal CI ran successfully, the open one is blocked on one Windows job that is not started but all other jobs succeeded thus CI can be considered green I think.

yanesca reviewed (approved):
I have reviewed the rebase and the last commit.
